node/network/bridge/src/lib.rs

```diff
@@ -0,0 +1,999 @@
+			NetworkBridgeMessage::ReportPeer(peer, rep) => Action::ReportPeer(peer, rep),
+			NetworkBridgeMessage::SendMessage(peers, protocol, message)
+				=> Action::SendMessage(peers, protocol, message),
+		},
+		Err(e) => {
+			log::warn!("Shutting down Network Bridge due to error {:?}", e);
+			Action::Abort
+		}
+	}
+}
+
+fn action_from_network_message(event: Option<NetworkEvent>) -> Option<Action> {
+	match event {
+		None => {
+			log::info!("Shutting down Network Bridge: underlying event stream concluded");
+			Some(Action::Abort)
+		}
+		Some(NetworkEvent::Dht(_)) => None,
+		Some(NetworkEvent::NotificationStreamOpened { remote, engine_id, role }) => {
+			if engine_id == POLKADOT_ENGINE_ID {
+				Some(Action::PeerConnected(remote, role))
+			} else {
+				None
+			}
+		}
+		Some(NetworkEvent::NotificationStreamClosed { remote, engine_id }) => {
+			if engine_id == POLKADOT_ENGINE_ID {
+				Some(Action::PeerDisconnected(remote))
+			} else {
+				None
+			}
+		}
+		Some(NetworkEvent::NotificationsReceived { remote, messages }) => {
+			let v: Result<Vec<_>, _> = messages.iter()
+				.filter(|(engine_id, _)| engine_id == &POLKADOT_ENGINE_ID)
+				.map(|(_, msg_bytes)| WireMessage::decode(&mut msg_bytes.as_ref()))
+				.collect();
+
+			match v {
+				Err(_) => Some(Action::ReportPeer(remote, MALFORMED_MESSAGE_COST)),
+				Ok(v) => if v.is_empty() {
+					None
+				} else {
+					Some(Action::PeerMessages(remote, v))
+				}
+			}
+		}
+	}
+}
+
+fn construct_view(live_heads: &[Hash]) -> View {
+	View(live_heads.iter().rev().take(MAX_VIEW_HEADS).cloned().collect())
+}
+
+async fn dispatch_update_to_all(
+	update: NetworkBridgeEvent,
+	event_producers: impl IntoIterator<Item=&fn(NetworkBridgeEvent) -> AllMessages>,
+	ctx: &mut impl SubsystemContext<Message=NetworkBridgeMessage>,
+) -> polkadot_subsystem::SubsystemResult<()> {
+	// collect messages here to avoid the borrow lasting across await boundary.
+	let messages: Vec<_> = event_producers.into_iter()
+		.map(|producer| producer(update.clone()))
+		.collect();
+
+	ctx.send_messages(messages).await
+}
+
+async fn update_view(
+	peers: &HashMap<PeerId, PeerData>,
+	live_heads: &[Hash],
+	net: &mut impl Network,
+	local_view: &mut View,
+) -> SubsystemResult<Option<NetworkBridgeEvent>> {
+	let new_view = construct_view(live_heads);
+	if *local_view == new_view { return Ok(None) }
+	*local_view = new_view.clone();
+
+	let message = WireMessage::ViewUpdate(new_view.clone()).encode();
+
+	let notifications = peers.keys().cloned()
+		.map(move |peer| Ok(NetworkAction::WriteNotification(peer, message.clone())));
+
+	net.action_sink().send_all(&mut stream::iter(notifications)).await?;
+
+	Ok(Some(NetworkBridgeEvent::OurViewChange(local_view.clone())))
+}
+
+async fn run_network<N: Network>(
+	mut net: N,
+	mut ctx: impl SubsystemContext<Message=NetworkBridgeMessage>,
+	action_inspector: impl Fn(&Action), // side-channel for tests to inspect internals
```

Comment:
I am indecisive whether we should have something as prominent as a function argument just for testing purposes (code smell).

Could one not inject a wrapped `ctx` and `event_stream` `Stream` in the unit test?

Reply:
Yeah, that is already done. Bu we'd then have an implicit dependence on the order that `run_network` processes messages from those things, which due to asynchrony is not possible. We need this action inspector to receive updates on what the `run_network` function has actually done.

The test harness uses `UnboundedSender` which can only tell you when something has been written, not when it has been read. The alternative would be to write a `Sink` implementation that only finishes polling once the message has been read.

Reply:
> The alternative would be to write a Sink implementation that only finishes polling once the message has been read.

I did this, so now all messages going into the subsystem are synchronous. However, messages coming out of the subsystem still use an unbounded channel so we don't have to process everything coming out, only the messages we need to test.